spec/datadog/profiling/stack_recorder_spec.rb

```diff
@@ -393,13 +397,67 @@ def sample_allocation(obj, _unused = 0)
           samples.select { |s| s.values[:'heap-live-samples'] == 0 }
         end
 
-        it 'include the stack and sample counts for the objects still left alive' do
-          # We sample from 2 distinct locations
-          expect(heap_samples.size).to eq(2)
+        matcher :approximate_age_based_on_allocation_order do |allocation_order|
+          # To prevent flakiness if other GCs other than the requested ones occur
+          # allow some margin in the matching. We do not expect this test to be
+          # particularly memory heavy so at most we expect one unexpected GC.
+          flaky_margin = 1
+
+          match do |gc_gen_age|
+            expected_age = num_allocations - allocation_order
+            gc_gen_age.between?(expected_age, expected_age + flaky_margin)
+          end
+          description do
+            expected_age = num_allocations - allocation_order
+            "be between #{expected_age} and #{expected_age + flaky_margin}"
+          end
+        end
+
+        it 'include the stack, sample counts and labels for the objects still left alive' do
+          # There should be 3 different allocation class labels so we expect 3 different heap samples
+          expect(heap_samples.size).to eq(3)
+
+          string_sample = heap_samples.find { |s| s.labels[:'allocation class'] == 'String' }
+          expect(string_sample).not_to be nil
```

Comment:
Minor: Since this testcase is already quite complex, I'd suggest removing this + the other matching assertions.

Anyway, if this happens there's other things that would fail, so at least we reduce a bit the complexity :)